### ISSUE TITLE
Incremental layout in 2020

### ISSUE THREAD

**Issue by SimonSapin** [servo/servo#25168]
The inputs of rendering includes the DOM tree, stylesheets, form data, input state (`:hover`, …), scroll positions, etc. When any of these inputs changes, we need to re-render a push a new bunch of pixels to the screen. It order to make this cheap, we want to reuse as much as possible of the previous rendering and skip re-computing intermediate results that haven’t changed. For layout in particular, this means reusing parts of the CSS box tree and fragment tree.

At the same time, we want fine-grained parallelism during layout to take advantage of available CPU cores. Avoiding data races requires either avoiding or synchronizing any shared mutability.

During our explorations last spring, I settled on trying the following approach to reconcile these goals:

* The box tree and fragment tree are each thread-safe and made of atomically reference-counted nodes.

* Nodes are *mostly* read-only after construction. Some pieces of data may be in `AtomicRefCell`s (if the algorithm structure is such that concurrent access would be a bug) or `Mutex`es or `RwLock`s, but most data can be read without synchronization.

* The trees are [persistent](https://en.wikipedia.org/wiki/Persistent_data_structure): updates are done by creating a new tree that share existing nodes for some sub-trees. "Modifying" a single node involves creating new nodes for it and its ancestors up to the root. This overhead is hopefully viable since HTML trees (which box trees and fragment trees are based on) tend not to be very deep, although they can get wide.

* Currently the child nodes of a given node are kept in a `Vec`. So creating a new node that is similar except with one different child node requires creating a new `Vec`, which takes O(n) time where n is the number of direct child nodes. It’s not uncommon for that number to grow large, so switching to [`im::Vector`](https://docs.rs/im/14.0.0/im/vector/enum.Vector.html) could help: `clone` is O(1) and `get_mut` is O(log(n)).

I think we could start with some conservative reuse in cases where we can easily determined that a subtree definitely hasn’t changed, and make it more fine-grained over time.

@nox, you mentioned an idea for reuse around block-in-inline splits?

**Comment by nox** [comment]
Yes, basically my idea lets us keep the children of a `display: contents` element in the box stored in that element itself, same for split inline boxes. That requires quite heavy changes to the DOM traversal code (but those changes would be local to it), and me figuring out the best way to represent a split inline box.

The end result of what I have in mind is that we incremental box generation would then be able to start from a list of root elements which were modified since last layout, and parents would only need to be reconstructed in a limited set of circumstances (for example, a previously-unsplit inline box in which a block box is inserted would trigger reconstructing its parent boxes until the nearest enclosing block box, but further edits to the split inline box wouldn't reconstruct parents anymore), AFAIK that gives us amortised `O(n)` incremental box construction, where `n` is the amount of nodes which changed.

**Comment by SimonSapin** [comment]
> keep the children of a `display: contents` element in the box stored in that element itself, same for split inline boxes

I don’t understand this part, could you rephrase?

**Comment by nox** [comment]
Some things straight out of my mind:

An element that generates an unsplit inline-level box would be handled just as currently, storing an `Arc<AtomicRefCell<InlineLevelBox>>` as its layout box, and when its parent constructs its own box, it will do the following:

* if the parent generates a block box, store the `Arc<AtomicRefCell<InlineLevelBox>>` as part of an anonymous inline formatting context.
* if the parent generates an inline box, store the `Arc<AtomicRefCell<InlineLevelBox>>` as a simple child of the parent's inline box.

Note how the inline-level child generates exactly one `Arc<AtomicRefCell<_>>` which can then be incrementally modified later without changing any parent inline formatting context or inline box.

On the other end, if that eleement generates a split inline-level box, we would store a value of a type like this instead:

```rust
struct SplitInlineLevelBox {
    leading_inline_level_box: Arc<AtomicRefCell<InlineLevelBox>>,
    inner_block_level_box: Arc<AtomicRefCell<BlockLevelBox>>,
    trailing_inline_level_box: Arc<AtomicRefCell<InlineLevelBox>>,
}
```

And on creating the box for the parent, two different things can happen again:

Either the parent is a block box, and we need to do three things:
 * append `leading_inline_level_box` to the current ongoing inline formatting context and end it
 * append the block-level box `inner_block_level_box` just as we would currently
 * start a new inline formatting context and put `trailing_inline_level_box` in it

The resulting layout box once every child has been handled is an `Arc<AtomicRefCell<BlockLevelBox>>`.

Or the parent is an inline-level box, and we need to do three things:
* wrap `leading_inline_level_box` in its own inline-level box with the parent's style
* wrap `trailing_inline_level_box` in its own inline-level box with the parent's style
* create a `SplitInlineLevelBox` with the two wrapped inline-level boxes, passing `inner_block_level_box` as is.

The resulting layout box is a `SplitInlineLevelBox`, and that box's parent box will need to take care of putting the leading and trailing inline-level boxes in the appropriate inline formatting contexts as usual, but no change to the leading and trailing inline-level boxes will require rebuilding the parent box, once again.

**Comment by pcwalton** [comment]
Do we actually need reference counting here?

When we go down the box tree to perform layout, we could keep the old render tree from the previous layout around and follow the links down through clean nodes with dirty descendants. Then when we get to a clean node with a clean descendant subtree, we just extract the old render subtree and use that instead of recreating it. Everything remains singly-owned at all times.

**Comment by nox** [comment]
We need reference counting whether or not there is incremental layout going on. Each box points to its children boxes, and each DOM element points to its layout box(es) too.

**Comment by pcwalton** [comment]
What about using `Arc::make_mut()` to perform persistent in-place updates of nodes? I noticed we already use it in a couple of places in layout. This has some overhead, but hopefully it won't be too high. Could even try using hardware lock elision.

**Comment by nox** [comment]
We still own those from 2 different places at once, so `Arc::make_mut` won't help us. I really doubt removing the refcounting is important here.

**Comment by nox** [comment]
So when you have 2 references to the same layout tree box from its DOM element and from its parent layout tree box, you have two choices.

Either you call `Arc::make_mut` on the DOM element's reference, which means the parent's reference is now wrong.

Or you call `Arc::make_mut` on the parent's reference, which means you have been calling `Arc::make_mut` recursively from the root of the layout tree, somehow managing to correlate it fully with the DOM tree it was generated from. I don't see how you can do that, and I don't automatically believe those recursive `Arc::make_mut` calls help either readability or performance.

**Comment by pcwalton** [comment]
> I don't automatically believe those recursive Arc::make_mut calls help either readability or performance.

As compared to mutex locks around every box?

**Comment by nox** [comment]
I don't think anyone of us expect to use R/W locks, just `AtomicRefCell` values.

**Comment by nox** [comment]
Note also that `Arc::make_mut` in the best case does just as much work as `AtomicRefCell::borrow_mut`.

**Comment by pcwalton** [comment]
You still think pervasive use of `AtomicRefCell` is feasible? (I don't disagree, by the way, but I was under the impression that for some reason you and Simon thought it was infeasible.)

**Comment by nox** [comment]
AFAIK no, we never said it was infeasible. [My comment](https://github.com/servo/servo/issues/25168#issuecomment-562616299) just before yours mentions `AtomicRefCell`.

**Comment by pcwalton** [comment]
Well, that's good news then :) My understanding per jdm was that there were doubts about it, but if not then carry on. `AtomicRefCell` seems like a good solution.

**Comment by pcwalton** [comment]
As a potential backup plan if `AtomicRefCell` is too slow, I propose something similar to https://bugzilla.mozilla.org/show_bug.cgi?id=1335941#c8 — make a special token that allows using unsynchronized writes to borrow `AtomicRefCell`s. We can use this token to perform sequential incremental reflow if we decide based on heuristics that parallel layout isn't worth it.

Another interesting avenue to consider is lock elision/hardware transactional memory. I could imagine something like "if hardware lock elision is available, then always do layout in parallel; otherwise, sometimes perform sequential incremental layout based on heuristics."

**Comment by pcwalton** [comment]
Started a branch for `AtomicRefCell`: https://github.com/pcwalton/servo/tree/layout-2020-atomic-refcell

**Comment by nox** [comment]
So are you working on this Patrick? The *last* thing I want is to duplicate efforts.

**Comment by nox** [comment]
A very first step towards incremental layout would be to skip box tree construction when none of `position`/`display`/`float` properties changed.

We need various things for that:
 1. Check that the style damage system already has bits for that, and if it doesn't, change it.
 2. Wrap the `ServoArc<ComputedValues>` from the various box tree types into its own `AtomicRefCell<_>` so that we can update those without trashing the entire box tree.
 3. Change box tree construction so that elements point to *all* the boxes they generated, even split inline elements.
 4. Don't rebuild the box tree when the damage system says it won't have changes, and instead update the style values in the boxes.

**Comment by nox** [comment]
(Also of course, we can't skip it if the document tree itself changed, e.g. when elements were added or removed.)

**Comment by pcwalton** [comment]
> So are you working on this Patrick? The _last_ thing I want is to duplicate efforts.

I'm not really intending to work in depth on incremental layout right now. I did #25957 primarily for performance testing. I submitted it because the perf results are positive and it's a fair bit of work so I wanted to save y'all the time of having to make those changes :)

**Comment by pcwalton** [comment]
> A very first step towards incremental layout would be to skip box tree construction when none of `position`/`display`/`float` properties changed.

Another good one is to skip layout entirely if only paint-relevant properties have changed. This covers a *lot* of common situations; e.g. mousing over links and selecting text.

**Comment by nox** [comment]
So I've looked into the current damage system.

The stuff in `style::restyle_damage` is pretty straightforwards: each property define what kind of damage it induces, and then all the various damages are checked to decide what should be reran and what should be reused.

The problem is that this system is really dependent on how layout is done, and thus the different categories of damages correspond to layout 2013 so much that I don't think it can be leveraged for layout 2020.

Do we want a new damage system that is specific to which layout system we are using instead? Like, does `ServoRestyleDamage::BUBBLE_ISIZES` make sense for layout 2020?

**Comment by pcwalton** [comment]
I agree with that; let's makes a new damage system. I don't have a lot of love for the old one anyway.

**Comment by pcwalton** [comment]
@nox Do you envision needing more `AtomicRefCell`s in layout past the ones I introduced in #25957?

**Comment by nox** [comment]
Good question! One thing I'm not sure about is how we update the styles of the trees we keep to regenerate the later trees that we don't keep. Do we just mutably borrow each box and fragment to change their style field, or do we store the `Arc<ComputedValues>` itself in an `Arc<AtomicRefCell<_>>` so we don't have to traverse trees just to update those? `Arc<AtomicRefCell<Arc<ComputedValues>>>` sounds a bit too much.

**Comment by nox** [comment]
`ArcSwap<Arc<ComputedValues>>` is also something we should look at if we go this way.

**Comment by nox** [comment]
In our Mako files, `servo_restyle_damage="rebuild_and_reflow"` means the box tree should change in layout 2020. We can reuse that.

**Comment by nox** [comment]
And `"reflow"` is more or less what corresponds to things that need to change the fragment tree.

**Comment by mrobinson** [comment]
Here's our current plan for how to proceed with incremental layout:

### Cache fragment tree layout and choose proper update point

There is already a small amount of incremental layout support in Servo, but two things are missing:

 - Caching of fragment tree results in the box tree
 - Proper selection of box tree update point (this currently breaks for absolutes)

Fragment tree results are currently cached in the box tree, but never reused.
When (2) is fixed here, we will be able to use that cache in more situations.

### Incrementally updating the box and fragment tree

 - Damage detection in style. This is likely necessary for all layout modes and
   will need incremental refinement.

 - Updating stylo so that various types of damage can be expressed to layout or
   alternatively, exposing a generic update mechanism from stylo.

 - Separating out layout phases so that only
   the parts of layout that are necessary are rebuilt. This requires looking at
   different layout modes to see how this process can be split up and what things
   need to be cached into the box tree.

 - Figuring out how to divide up phases of layout an stitch together old and
   new box and fragment trees during partial tree updates.

**Comment by coding-joedow** [comment]
# 1. Background

In Servo layout 2020, the fragment tree as the layout result, and the box tree as the layout input, are two separate trees. When the Dom tree is dirty, it’s necessary to update the box tree to follow the changes in the Dom tree. Then, the updated box tree is used as the input, re-layout, and output the new fragment tree. 

In the current layout 2020 design, because no direct link exists between the box tree and the fragment tree, there is no any information about the last layout result can be obtained in the box tree. Therefore, all work is completely re-executed in each layout.

To support incremental layouts and reduce unnecessary work that needs to be re-executed, some layout results or even intermediate layout results need to be stored on the box tree, which will be provided as input for the next layout. Accordingly, incremental updates also need to be supported on the box tree. Then the last layout result can be reused as much as possible.
# 2. Current Status

Currently, the box tree is almost always completely rebuilt. Under the current update policy of the box tree, only valid candidates as the update point of the box tree are the absolute positioning ancestor nodes of the dirty root, which outside display mode is kept unchanged. Then, traversing down from the update point and rebuild all the boxes. This strategy is too conservative:
- On a real page, the number of absolute positioning elements is very small, so it is more common for the root element to be selected as the update point.
- Even if only the sub-trees under the update point are rebuilt, there are still a considerable number of nodes that may not need to be rebuilt.

There are some pain points in incrementally updating the box tree, which I guess is the reason why currently only the absolute positioning elements are selected as the candidate set of the box tree update points:
- It’s not easy to determine whether a node’s boxes needs to rebuilt, which depends not only on its own style changes, but also on changes in the context in which it is located, such as changes on the parent, child, or even descendant nodes;
- With the current box tree representation, it is difficult to incrementally check whether the propagated data changes, such as text-decoration-line and contains float.

Thus it is safe enough for the out of flow absolutely positioned nodes to be candidates for box tree update points.
# 3. Design scheme

Next is a design proposal to improve the current box tree update, including: detecting box tree damage, expanding the update point candidate set, and supporting node-by-node update of the box tree instead of fully reconstruction of the entire sub-tree, so as to reduce the number of reconstructed boxes and indirectly reduce the redone works for incremental layouts.
## 3.1 Detecting the box tree damage

Detecting the node’s damage is divided into three steps:
1) Check the damage caused by the node itself style changes;
2) Propagating down the damage from the ancestor node;
3) Propagating up the damage from the descendants;
### 3.1.1  Style Changes

Currently, the stylo crate supports to detect the style difference’s impact on the subsequent phases of the reflow during the style recalculation phase, which is called **_```RestyleDamage```_**. However, the detection mechanism is designed for layout 2013. Because the layout 2020’s implementation is significantly different from layout 2013, this restyle damage detection mechanism needs to be updated for layout 2020.

The **_```RestyleDamage```_** has a bit indicating whether a node needs to rebuild its boxes, as shown in the following figure. Whether the bit should be marked depends on the difference between the attributes of the old style and the new style of the node. For layout 2020, the attributes set need to be adjusted.
```rust
bitflags! {
    pub struct ServoRestyleDamage: u8 {
        // ...
        const RECONSTRUCT_FLOW = 0x80;
        // ...
    }
}
```
### 3.1.2  Top-down propagation

Then considering how the change in the ancestor node affects whether the node’s boxes need to be rebuilt:
- a. Conservatively, when a node needs to rebuild a box, all descendants of the node need to be rebuilt. This is also the current implementation of servo and blink, and is the simplest way.
- b. Or propagating the damage down on demands:
	- To balance performance and complexity, if the node's primary box needs to be rebuilt, and the node's primary box holds the ```IFC``` or its anonymous boxes hold the ```IFC```, all the inline descendant elements participating in these ```IFC```, their boxes need to be rebuilt. Because the construction of ```IFC``` need to collect all inline level descendant elements and are split by the block level descendant elements, it is difficult to incrementally determine which inline level descendant element in an ```IFC``` needs to be rebuilt.
	- If the primary box of a node needs to be rebuilt, the pseudo-element boxes and list marker box also need to be rebuilt, and vice versa.
	- If the node creates a new formatting context, all of its children’s boxes need to be rebuilt. For example, if a node is changed from participating in or creating a ```BFC``` to creating a flex container, then its child nodes need to be changed from participating in the ```BFC``` to creating a new ```BFC```.
	- If the node itself is a table container or it has an anonymous table, both all descendant nodes under the table and this table container itself always need to be rebuilt. Under the current servo layout implementation, the layout data of a node with the table-related internal display mode, is not associated with any box. As a result, it is difficult to determine whether such a node needs to be rebuilt.

I prefer to choose b to maximize the chances of the last layout result being reused during next layout. This step can be implemented during the top-down process of the incremental box tree updating.
### 3.1.2  Bottom-up propagation

Finally, considering how the changes of the descendant node affect whether the node itself needs to be rebuilt. There are two strategies:
- a. Conservatively, the node always needs to be rebuilt unconditionally when one of its child needs to be rebuilt, which is true until the selected box tree update point. In fact, this is the implementation of layout 2013, and is a relatively simple way.
- b. Or propagating up the damage on demands:
	- If the outside display mode of the node changes, or the node is changed between in-flow and out-flow mode, or an inline level node required to be rebuilt due to other reason, the ```IFC``` that the node participates in needs to be rebuilt or constructed. Propagating up the damage to the ```IFC``` owner, which is the first exclusive block level node.
- If a child node is added or deleted, the parent node always needs to be rebuilt.

I prefer to choose b to maximize the chances of the last layout result being reused during next layout. This process can be implemented either during the style recalculation process’ the post-order traversal phase, as in layout 2013, or during the box tree update process. I prefer to do it during the post-order traversal of the style recalculation process.

Specially, the combination of 3.1.2-a and 3.1.3-a means that all descendants of the update point need to be rebuilt.
## 3.2   Expanding the candidate set of update points

The valid candidate set can be extended to nodes that satisfy the following conditions: the node creates an independent formatting context and its display mode remains unchanged, which means:
- The influence of the float element is limited within the ```BFC```, so there is no need to check whether the ancestor of the node that created the ```BFC``` contains float.
- The box sub-tree reconstruction inside the independent formatting context does not affect the outside of the formatting context. Therefore, there is no need to propagate up the box tree damage.

Then, considering the data that are propagated from the top down, such as the text-decoration:
- Backtracking up from the box tree update point to find the absolute positioning element or root element, then, accumulating down the propagated box tree data until the box tree update point.
- Alternatively, the accumulated propagated box tree data is directly stored on each box, which can be reused at next box tree update but requires additional memory space.
## 3.3   Updating the box tree node by node based on damage

This work is divided into box repair and box reconstruction. The box repair only updates the information carried by the box, including the computed style, propagated box tree data, contains float, and children boxes list, etc.

The following is a preliminary algorithm for updating the box tree node by node. Traverse the Dom tree top-down from the update point:
- **A.** If the node is marked as damaged by ```RECONSTRUCT_FLOW```, go to step **B**. Otherwise, go to step **C**.
- **B.** Rebuild the node’s boxes, and traverse all child nodes and inline level descendants.
	- **a.** Check whether its own damage needs to be propagated to descendant nodes. For details, see section **3.1.2**. If yes or that this descendant node has already been marked damage, go to step **B-b**. Otherwise, go to step **B-c**. The boxes of all pseudo-elements and list marker are forced to be rebuilt.
	- **b.** Reconstruct boxes for this descendant according to the existing box tree construction implementation.
	- **c.** Repair boxes for the this descendant:
		- i. If it is block level node and participates in the ```BFC```, the ```BlockLevelJob``` with ```BlockLevelCreator::RepairedBox``` type is generated and added to the job queue.
		-  ii. For the inline level node, the reconstruction of its relevant ```IFC``` is skipped, but the traversal of its descendant elements is maintained, so as to search for the block level descendants which can split the ```IFC```. If such descendants are found, go to step **A** to process them.
		- iii. If this node is a flex item or grid item, the ```ModernContainerJob``` with the ```ModernContainerJob::RepairedBox``` type are generated and added to the job queue
	- **d.** Drain the job queue to process each job. Go to step **B** or step **C** based on the job type, and output a new box or original box reference for the relevant node.
	- **e.** Then, build a new box for this node with a new child boxes list.
- **C.** Repair the node’s boxes, and traverse all child nodes and descendant inline nodes.
	- **a.** If the traversed node is marked as damaged by ```RECONSTRUCT_FLOW```, go to step **B-b**. Otherwise, go to step **B-c**.
	- **b.** After the preceding steps are complete, drain the job queue to process each job. Go to step **B** or step **C** based on the job type, and output a new box or original box reference for the relevant node.
	- **c.** Then, replace the old child boxes list with the new collected one, and update the other information for the node’s primary box.

The following list the types of box tree repair jobs that may need to be added:
```rust
enum BlockLevelCreator {
    SameFormattingContextBlock(IntermediateBlockContainer),
    // ...
    RepairedBox(...)
}

enum ModernContainerJob {
	ElementOrPseudoElement(...),
	TextRuns(...),
	RepairedBox(...)
}
```

In the initial design, the following ```structs``` or ```enums``` maybe need to be added to the layout 2020:
- **```BlockContainerRepairer```**: It replaces ```BlockContainerBuilder``` to repair the boxes.

Maybe need to modify the following ```struct``` or ```enum```:
- **```DomTraversal```**: Modify its implementation to support the damage propagation during post-order bottom-top traversal of style calculation.
- **```BlockLevelCreator```**: Added a new enumerated value to support the box repair.
- **```BlockLevelJob```**: Modify the implementation to support the box repair.
- **```ModernContainerJob```**: Added a new enumerated value to support the box repair.
- **```ModernItemKind```**: Added a new enumerated value to support the box repair.
- **```ModernContainerBuilder```**: Modified implementation to support box repair

**Comment by coding-joedow** [comment]
> Here's our current plan for how to proceed with incremental layout:
> 
> ### Cache fragment tree layout and choose proper update point
> There is already a small amount of incremental layout support in Servo, but two things are missing:
> 
> * Caching of fragment tree results in the box tree
> * Proper selection of box tree update point (this currently breaks for absolutes)
> 
> Fragment tree results are currently cached in the box tree, but never reused. When (2) is fixed here, we will be able to use that cache in more situations.
> 
> ### Incrementally updating the box and fragment tree
> * Damage detection in style. This is likely necessary for all layout modes and
>   will need incremental refinement.
> * Updating stylo so that various types of damage can be expressed to layout or
>   alternatively, exposing a generic update mechanism from stylo.
> * Separating out layout phases so that only
>   the parts of layout that are necessary are rebuilt. This requires looking at
>   different layout modes to see how this process can be split up and what things
>   need to be cached into the box tree.
> * Figuring out how to divide up phases of layout an stitch together old and
>   new box and fragment trees during partial tree updates.


About “Proper selection of box tree update point”, i have a detailed proposal. More precisely, it's about incremental box tree updates. Perhaps there are some over-designed or unconsidered areas in this proposal, and I look forward to your reply.

# 1. Background

In Servo layout 2020, the fragment tree as the layout result, and the box tree as the layout input, are two separate trees. When the Dom tree is dirty, it’s necessary to update the box tree to follow the changes in the Dom tree. Then, the updated box tree is used as the input, re-layout, and output the new fragment tree. 

In the current layout 2020 design, because no direct link exists between the box tree and the fragment tree, there is no any information about the last layout result can be obtained in the box tree. Therefore, all work is completely re-executed in each layout.

To support incremental layouts and reduce unnecessary work that needs to be re-executed, some layout results or even intermediate layout results need to be stored on the box tree, which will be provided as input for the next layout. Accordingly, incremental updates also need to be supported on the box tree. Then the last layout result can be reused as much as possible.
# 2. Current Status

Currently, the box tree is almost always completely rebuilt. Under the current update policy of the box tree, only valid candidates as the update point of the box tree are the absolute positioning ancestor nodes of the dirty root, which outside display mode is kept unchanged. Then, traversing down from the update point and rebuild all the boxes. This strategy is too conservative:
- On a real page, the number of absolute positioning elements is very small, so it is more common for the root element to be selected as the update point.
- Even if only the sub-trees under the update point are rebuilt, there are still a considerable number of nodes that may not need to be rebuilt.

There are some pain points in incrementally updating the box tree, which I guess is the reason why currently only the absolute positioning elements are selected as the candidate set of the box tree update points:
- It’s not easy to determine whether a node’s boxes needs to rebuilt, which depends not only on its own style changes, but also on changes in the context in which it is located, such as changes on the parent, child, or even descendant nodes;
- With the current box tree representation, it is difficult to incrementally check whether the propagated data changes, such as text-decoration-line and contains float.

Thus it is safe enough for the out of flow absolutely positioned nodes to be candidates for box tree update points.
# 3. Design scheme

Next is a design proposal to improve the current box tree update, including: detecting box tree damage, expanding the update point candidate set, and supporting node-by-node update of the box tree instead of fully reconstruction of the entire sub-tree, so as to reduce the number of reconstructed boxes and indirectly reduce the redone works for incremental layouts.
## 3.1 Detecting the box tree damage

Detecting the node’s damage is divided into three steps:
1) Check the damage caused by the node itself style changes;
2) Propagating down the damage from the ancestor node;
3) Propagating up the damage from the descendants;
### 3.1.1  Style Changes

Currently, the stylo crate supports to detect the style difference’s impact on the subsequent phases of the reflow during the style recalculation phase, which is called **_```RestyleDamage```_**. However, the detection mechanism is designed for layout 2013. Because the layout 2020’s implementation is significantly different from layout 2013, this restyle damage detection mechanism needs to be updated for layout 2020.

The **_```RestyleDamage```_** has a bit indicating whether a node needs to rebuild its boxes, as shown in the following figure. Whether the bit should be marked depends on the difference between the attributes of the old style and the new style of the node. For layout 2020, the attributes set need to be adjusted.
```rust
bitflags! {
    pub struct ServoRestyleDamage: u8 {
        // ...
        const RECONSTRUCT_FLOW = 0x80;
        // ...
    }
}
```
### 3.1.2  Top-down propagation

Then considering how the change in the ancestor node affects whether the node’s boxes need to be rebuilt:
- a. Conservatively, when a node needs to rebuild a box, all descendants of the node need to be rebuilt. This is also the current implementation of servo and blink, and is the simplest way.
- b. Or propagating the damage down on demands:
	- To balance performance and complexity, if the node's primary box needs to be rebuilt, and the node's primary box holds the ```IFC``` or its anonymous boxes hold the ```IFC```, all the inline descendant elements participating in these ```IFC```, their boxes need to be rebuilt. Because the construction of ```IFC``` need to collect all inline level descendant elements and are split by the block level descendant elements, it is difficult to incrementally determine which inline level descendant element in an ```IFC``` needs to be rebuilt.
	- If the primary box of a node needs to be rebuilt, the pseudo-element boxes and list marker box also need to be rebuilt, and vice versa.
	- If the node creates a new formatting context, all of its children’s boxes need to be rebuilt. For example, if a node is changed from participating in or creating a ```BFC``` to creating a flex container, then its child nodes need to be changed from participating in the ```BFC``` to creating a new ```BFC```.
	- If the node itself is a table container or it has an anonymous table, both all descendant nodes under the table and this table container itself always need to be rebuilt. Under the current servo layout implementation, the layout data of a node with the table-related internal display mode, is not associated with any box. As a result, it is difficult to determine whether such a node needs to be rebuilt.

I prefer to choose b to maximize the chances of the last layout result being reused during next layout. This step can be implemented during the top-down process of the incremental box tree updating.
### 3.1.2  Bottom-up propagation

Finally, considering how the changes of the descendant node affect whether the node itself needs to be rebuilt. There are two strategies:
- a. Conservatively, the node always needs to be rebuilt unconditionally when one of its child needs to be rebuilt, which is true until the selected box tree update point. In fact, this is the implementation of layout 2013, and is a relatively simple way.
- b. Or propagating up the damage on demands:
	- If the outside display mode of the node changes, or the node is changed between in-flow and out-flow mode, or an inline level node required to be rebuilt due to other reason, the ```IFC``` that the node participates in needs to be rebuilt or constructed. Propagating up the damage to the ```IFC``` owner, which is the first exclusive block level node.
- If a child node is added or deleted, the parent node always needs to be rebuilt.

I prefer to choose b to maximize the chances of the last layout result being reused during next layout. This process can be implemented either during the style recalculation process’ the post-order traversal phase, as in layout 2013, or during the box tree update process. I prefer to do it during the post-order traversal of the style recalculation process.

Specially, the combination of 3.1.2-a and 3.1.3-a means that all descendants of the update point need to be rebuilt.
## 3.2   Expanding the candidate set of update points

The valid candidate set can be extended to nodes that satisfy the following conditions: the node creates an independent formatting context and its display mode remains unchanged, which means:
- The influence of the float element is limited within the ```BFC```, so there is no need to check whether the ancestor of the node that created the ```BFC``` contains float.
- The box sub-tree reconstruction inside the independent formatting context does not affect the outside of the formatting context. Therefore, there is no need to propagate up the box tree damage.

Then, considering the data that are propagated from the top down, such as the text-decoration:
- Backtracking up from the box tree update point to find the absolute positioning element or root element, then, accumulating down the propagated box tree data until the box tree update point.
- Alternatively, the accumulated propagated box tree data is directly stored on each box, which can be reused at next box tree update but requires additional memory space.
## 3.3   Updating the box tree node by node based on damage

This work is divided into box repair and box reconstruction. The box repair only updates the information carried by the box, including the computed style, propagated box tree data, contains float, and children boxes list, etc.

The following is a preliminary algorithm for updating the box tree node by node. Traverse the Dom tree top-down from the update point:
- **A.** If the node is marked as damaged by ```RECONSTRUCT_FLOW```, go to step **B**. Otherwise, go to step **C**.
- **B.** Rebuild the node’s boxes, and traverse all child nodes and inline level descendants.
	- **a.** Check whether its own damage needs to be propagated to descendant nodes. For details, see section **3.1.2**. If yes or that this descendant node has already been marked damage, go to step **B-b**. Otherwise, go to step **B-c**. The boxes of all pseudo-elements and list marker are forced to be rebuilt.
	- **b.** Reconstruct boxes for this descendant according to the existing box tree construction implementation.
	- **c.** Repair boxes for the this descendant:
		- i. If it is block level node and participates in the ```BFC```, the ```BlockLevelJob``` with ```BlockLevelCreator::RepairedBox``` type is generated and added to the job queue.
		-  ii. For the inline level node, the reconstruction of its relevant ```IFC``` is skipped, but the traversal of its descendant elements is maintained, so as to search for the block level descendants which can split the ```IFC```. If such descendants are found, go to step **A** to process them.
		- iii. If this node is a flex item or grid item, the ```ModernContainerJob``` with the ```ModernContainerJob::RepairedBox``` type are generated and added to the job queue
	- **d.** Drain the job queue to process each job. Go to step **B** or step **C** based on the job type, and output a new box or original box reference for the relevant node.
	- **e.** Then, build a new box for this node with a new child boxes list.
- **C.** Repair the node’s boxes, and traverse all child nodes and descendant inline nodes.
	- **a.** If the traversed node is marked as damaged by ```RECONSTRUCT_FLOW```, go to step **B-b**. Otherwise, go to step **B-c**.
	- **b.** After the preceding steps are complete, drain the job queue to process each job. Go to step **B** or step **C** based on the job type, and output a new box or original box reference for the relevant node.
	- **c.** Then, replace the old child boxes list with the new collected one, and update the other information for the node’s primary box.

The following list the types of box tree repair jobs that may need to be added:
```rust
enum BlockLevelCreator {
    SameFormattingContextBlock(IntermediateBlockContainer),
    // ...
    RepairedBox(...)
}

enum ModernContainerJob {
	ElementOrPseudoElement(...),
	TextRuns(...),
	RepairedBox(...)
}
```

In the initial design, the following ```structs``` or ```enums``` maybe need to be added to the layout 2020:
- **```BlockContainerRepairer```**: It replaces ```BlockContainerBuilder``` to repair the boxes.

Maybe need to modify the following ```struct``` or ```enum```:
- **```DomTraversal```**: Modify its implementation to support the damage propagation during post-order bottom-top traversal of style calculation.
- **```BlockLevelCreator```**: Added a new enumerated value to support the box repair.
- **```BlockLevelJob```**: Modify the implementation to support the box repair.
- **```ModernContainerJob```**: Added a new enumerated value to support the box repair.
- **```ModernItemKind```**: Added a new enumerated value to support the box repair.
- **```ModernContainerBuilder```**: Modified implementation to support box repair

**Comment by shangguan1024** [comment]
Does supporting incremental layout affect the output display list? Currently, WebRender receives a full display list, and processing it in the SceneBuilder thread takes some time. Would it be feasible to implement incremental display list processing in WebRender?

**Comment by mrobinson** [comment]
> Does supporting incremental layout affect the output display list? Currently, WebRender receives a full display list, and processing it in the SceneBuilder thread takes some time. Would it be feasible to implement incremental display list processing in WebRender?

We currently don't support incremental display list updates, so the easiest thing to do initially would be to always generate a full display list (and stacking context tree) every time we relayout.

**Comment by mrobinson** [comment]
I think we can close this now as we have the beginnings of a decent incremental layout system, especially after #45561 lands (which is the first time we can reuse fragments in the tree above fragment tree damage).
